### PR TITLE
Output Schema

### LINKED ISSUE #918: Add `output_schema` to SAML.
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/homanp/superagent/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/homanp/superagent/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

If a agent yaml node includes `output_schema` we should add a parsing agent as the last step in a workflow. 
This parsing agent takes the `output_schema` content as it's prompt return a structured object. This agent could just be a simple call to OpenAI. 

Example SAML: 

```yaml
workflows:
  - superagent:
       name: test
       llm: gpt-3.5-turbo-0125
       prompt: Hello 
       output_schema: 
         - my_key: string
         - my_second_key: number
```

This should support common types such as `enum`, `number` and `string`. 

### Acknowledgements

- [X] My issue title is concise and descriptive.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by elisalimli**
## Summary

Fixes #918

## Depends on
#925 

## Test plan

Here's an example SAML that I tested to scrape structured data from Amazon.com

```yaml
workflows:
  - superagent:
      llm: gpt-3.5-turbo-1106
      name: Browser assistant
      intro: 👋 Hi there! How can I help search for answers on the internet.
      tools:
        - scraper:
            name: browser
            use_for: searching the internet
            metadata:
              apiKey: 9EA0OH7T2LXAXHPV3K036MN192XQZ2DRMP2XA9QC0KF2LBU9XCMXK8CJREIU5AVNOYPMHJD4Y2AEF8NR
      prompt: Use the browser to answer any questions
      output_schema:
        type: array
        items:
          type: object
          required:
            - id
            - name
          properties:
            id:
              type: integer
            product_name:
              type: integer
            product_price:
              type: string
```

```
  "input": "i need product name, product price, store name from https://www.amazon.com/s?k=apple+m3&crid=2CSLNASAYI715&sprefix=apple+mc%2Caps%2C376&ref=nb_sb_noss_2",
```

### CODE REVIEW

File: libs/superagent/prisma/migrations/20240328135202_add_output_schema_field_agents_table/migration.sql
Comment (homanp) [inline libs/superagent/prisma/migrations/20240328135202_add_output_schema_field_agents_table/migration.sql:1]: Why do we need to store this separately? It's already stored in the config?

**Review by homanp — COMMENTED** [review]
Left a comment @elisalimli

**Review by homanp — CHANGES_REQUESTED** [review]
@elisalimli this needs work with the REST API and SDKs as well. could you add that as well? On the workflow `invoke` function.

**Comment by elisalimli** [comment]
> @elisalimli this needs work with the REST API and SDKs as well. could you add that as well? On the workflow `invoke` function.

@homanp it will work on the SDK too. We just call our existing api_endpoints in SAML. So whatever works there will work on REST API/SDK

File: libs/superagent/prisma/migrations/20240328135202_add_output_schema_field_agents_table/migration.sql
Comment (elisalimli) [reply libs/superagent/prisma/migrations/20240328135202_add_output_schema_field_agents_table/migration.sql:1]: Where is it stored? Please point to sources if I missed any.

**Comment by homanp** [comment]
> > @elisalimli this needs work with the REST API and SDKs as well. could you add that as well? On the workflow `invoke` function.
> 
> @homanp it will work on the SDK too. We just call our existing api_endpoints in SAML. So whatever works there will work on REST API/SDK

I don't see any `outputSchema` key work invoking workflows in the REST API.

**Comment by elisalimli** [comment]
> > > @elisalimli this needs work with the REST API and SDKs as well. could you add that as well? On the workflow `invoke` function.
> > 
> > 
> > @homanp it will work on the SDK too. We just call our existing api_endpoints in SAML. So whatever works there will work on REST API/SDK
> 
> I don't see any `outputSchema` key work invoking workflows in the REST API.

Yes, we don't. I thought it's redundant to receive outputSchema every time. Instead we're storing in the database and read from there.

File: libs/superagent/app/api/workflow_configs/data_transformer.py [hunk 1]
```diff
@@ -175,6 +186,7 @@ async def _set_database_provider(self, datasource: dict):
                 f"Database provider is not set, using default provider - {database}"
             )
 
+        print(database, database_provider)
```
Comment (homanp) [inline libs/superagent/app/api/workflow_configs/data_transformer.py]: Remove

**Review by homanp — CHANGES_REQUESTED** [review]

**Review by homanp — CHANGES_REQUESTED** [review]
Screenshot 2024-03-29 at 11 38 52

**Review by homanp — APPROVED** [review]

**Pull request merged**